src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1075,7 +1075,28 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             }
 
             diagnostics.AddRangeAndFree(localDiagnostics);
-            var boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, type: declTypeOpt.TypeSymbol);
+
+            ImmutableArray<BoundExpression> dimensionsOpt;
+            if (typeSyntax is ArrayTypeSyntax arrayTypeSyntax && arrayTypeSyntax.RankSpecifiers.Count > 0)
```

Comment:
>typeSyntax is ArrayTypeSyntax arrayTypeSyntax [](start = 16, length = 45)

Do we need to worry about type syntax other than ArrayTypeSyntax? For example, a generic type with an array type as a type argument.

Reply:
In an array creation that would also be invalid. For now I'm just trying to capture cases where the syntax would be valid in an ArrayCreationExpression, but invalid in a variable declaration, as I feel these are probably the most common cases.

I think there a huge amount of scope to do more here. This is meant to be an initial pull request with the minimum changes required to be useful.

Reply:
> For now I'm just trying to capture cases where the syntax would be valid in an ArrayCreationExpression, but invalid in a variable declaration, as I feel these are probably the most common cases.

Why are these scenarios more interesting? What has changed in variable declarations around array type syntax?  


> I think there a huge amount of scope to do more here. This is meant to be an initial pull request with the minimum changes required to be useful.

I do not think I support this approach. It becomes very hard to make changes to features that has shipped due to backward compatibility considerations. This PR goes to master. It is not clear to me what is the final plan, how do we think it should look like at the end and what is the plan for us to get there before we ship. I'd rather have clarity about all the scenarios that we should handle here.  

---
In reply to: [251617370](https://github.com/dotnet/roslyn/pull/32843#discussion_r251617370) [](ancestors = 251617370)

Reply:
I believe we would have to put the member on a different node of the IOperation tree, so I view this as two seperate APIs.

I listed a large number of potential places where we could expose this information at the top of the pull request. I think we have to decide exactly which of them should go in the initial PR.

Reply:
I think that we should handle all types here, not just array types. Dig through any type and collect sizes from all array types within it.

---
In reply to: [251623089](https://github.com/dotnet/roslyn/pull/32843#discussion_r251623089) [](ancestors = 251623089,251617370)

Reply:
> Do we need to worry about type syntax other than ArrayTypeSyntax? For example, a generic type with an array type as a type argument.

This always ends up being parsed as a BinaryExpression, as when a generic type declaration cannot be parsed correctly, it seems we parse `<` as Less Than.

I can't think of any cases where we would parse an expression type as type syntax which would contain an invalid dimension other than an ArrayTypeSyntax.

Reply:
>This always ends up being parsed as a BinaryExpression, as when a generic type declaration cannot be parsed correctly, it seems we parse ```<``` as Less Than.

>I can't think of any cases where we would parse an expression type as type syntax which would contain an invalid dimension other than an ArrayTypeSyntax.

Can one create such a tree manually by using SyntaxFactory? Also, I might be wrong, but I think with the recent changes to the parser an element type of the array type could be a nullable array type with sizes as well. Also, since we are going to do the collection of misplaced sizes for other contexts too, and in those contexts parser's behavior around handling of ```<``` might be slightly different, or it might change in the future, I think it would be more robust to create a helper that doesn't make any assumption on how the parser works, simply visits the syntax (by using a syntax visitor) and collects all size expressions from array types in the syntax. 


---
In reply to: [253308012](https://github.com/dotnet/roslyn/pull/32843#discussion_r253308012) [](ancestors = 253308012)